adding the FileSaver.js library

felipesanches opened the pull request:
(used for prompting the user with a "Save as..." dialog when downloading ZIP UFO fonts)

davelab6 commented:
If I have a default download folder set to automatically download files
without bothering me, will this bother me with where to save them?

felipesanches commented:
I'm not sure. I guess it must respect your settings. The feature is implemented by using the browser's native file save feature, so I expect that the browser implementation will take care of this. 

But I may be wrong. We'll figure it out soon and I'll report here (or here: #410).

davelab6 commented:
Okay so no worries, I guess this allows js objects to get downloaded,
that's all
On 17 Apr 2015 6:23 pm, "Felipe Corrêa da Silva Sanches" <
notifications@github.com> wrote:

> I'm not sure. I guess it must respect your settings. The feature is
> implemented by using the browser's native file save feature, so I expect
> that the browser implementation will take care of this.
>
> But I may be wrong. We'll figure it out soon and I'll report here (or
> here: #410 <https://github.com/metapolator/metapolator/pull/410>).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/metapolator/metapolator/pull/411#issuecomment-94087782>
> .
>

graphicore commented on app/lib/project/MetapolatorProject.js:
```diff
@@ -20,6 +20,7 @@ define([
   , 'yaml'
   , 'io/zipUtil'
   , 'io/InMemory'
+  , 'filesaver'
```
You require it as a dependency, so you should receive it in the module closure function as an argument.

In this case you want to receive it under the name `saveAs`

graphicore replied:
This is important even if you are not going to use it for whatever odd reason, because the next module required hereafter will need to name it's argument also, and not using the right slot causes confusion.

graphicore commented on app/lib/browserConfig.js:
```diff
@@ -21,6 +21,7 @@ requirejs.config({
       , 'socketio': '../socket.io/socket.io'
       , 'EventEmitter': 'bower_components/event-emitter.js/dist/event-emitter'
       , 'jszip': 'bower_components/jszip/dist/jszip'
+      , 'filesaver': 'bower_components/file-saver.js/FileSaver'
```
It would be better to name the module here `saveAs` so it's more obvious what you require. This FileSaver module exports just one function called `saveAs`.

graphicore commented on app/lib/project/MetapolatorProject.js:
```diff
@@ -626,6 +627,10 @@ define([
     };
 
+    _p.saveAs = function(blob, filename) {
+        saveAs(blob, filename);
```
This is a no-go. As I understand it, `filesaver` is expected to declare `saveAs`. But that is entirely implicit. You should receive `saveAs` as argument of the module funnction. It's bad to declare a dependency and not to use it, anyways. So if you are not going to use a dependency directly, you should still give it a name in the module closure arguments.

Anyways, filesaver does this:
```js
define([], function() {
    return saveAs;
});
```

so you can do this:

```
define(['filesaver'], function(saveAs){ 
/* use saveAs() here */
})

```

graphicore commented on app/lib/project/MetapolatorProject.js:
```diff
@@ -626,6 +627,10 @@ define([
         }
     };
 
+    _p.saveAs = function(blob, filename) {
```
I don't think metapolatorProject is a good fit for this.  This is just "saving" data, it's not specific for a Metapolator Project at all. It may not even work within NodeJS, that might be the reason for the tests not passing at the moment.

Rather we can add it to `metapolatorStandAlone` and just export it from there for use from the UI.

If this is really not loading within NodeJS. we should even consider adding a dummy package that raises a NotImplentedError when trying to use the saveAs method or so.

graphicore replied:
Even if this would be a good place for the `saveAs` method, I think the implementation here should be done so:

```js
_p.saveAs = saveAs;
```

felipesanches commented:
Could you please verify if this last commit correctly addresses all of your concerns?

I'm still unsure about where to declare the saveAs method. It is still declared as a method of MetapolatorProject. Where should we move it to?

graphicore commented on config.js:
```diff
@@ -17,6 +17,7 @@ module.exports = {
       , 'es6/Reflect': 'bower_components/harmony-reflect/reflect'
       , 'codemirror': 'bower_components/codemirror'
       , 'jszip': 'bower_components/jszip/dist/jszip'
+      , 'filesaver': 'bower_components/file-saver.js/FileSaver'
```
Did you try to load this in NodeJS?

For me:

```
$ metapolator dev-js-shell
metapolator> require('filesaver')
Error: Evaluating /home/commander/Projekte/metapolator/vectorworks/app/lib/bower_components/file-saver.js/FileSaver.js as module "filesaver" failed with error: TypeError: Cannot read property 'document' of undefined
    at Function.req.load (/home/commander/Projekte/metapolator/vectorworks/node_modules/requirejs/bin/r.js:2529:23)
    at Function.req.get (/home/commander/Projekte/metapolator/vectorworks/node_modules/requirejs/bin/r.js:2479:29)
    at Object.localRequire [as require] (/home/commander/Projekte/metapolator/vectorworks/node_modules/requirejs/bin/r.js:1650:40)
    at requirejs (/home/commander/Projekte/metapolator/vectorworks/node_modules/requirejs/bin/r.js:2003:24)
    at repl:1:2
    at REPLServer.self.eval (repl.js:112:21)
    at Interface.<anonymous> (repl.js:239:12)
    at Interface.EventEmitter.emit (events.js:95:17)
    at Interface._onLine (readline.js:202:10)
    at Interface._line (readline.js:531:8)
```

Because this is not compatible with nodeJS. `document` is part of the Web API.

felipesanches replied:
I see. Well... saveAs will not be used at all by nodeJS. So how should we proceed, then?

graphicore replied:
See my last comment, it's all in there: https://github.com/metapolator/metapolator/pull/411#issuecomment-94572896 :-)

graphicore commented:
In https://github.com/metapolator/metapolator/blob/master/app/lib/metapolatorStandAlone.js

First import filesaver:
```js
require([
// ...
  , 'filesaver'
],
function (
// ...
  , saveAs
) {
// ...
```

Then export filesaver:

```js
// ... way down at the botttom
    exports.saveAs = saveAs;
});
```

In the ui code you can do now:
```js
$scope.data.stateless.saveAs(zipped_data, filename)
```

And remove `filesaver` from `metapolatorProject` and `config.js`

felipesanches commented:
OK, I'll do it. Thanks for the tip! :-)

graphicore commented:
OK, looks good. Also the tests pass again :-) If you can confirm that it works within the ui I merge this.

felipesanches commented:
yes. I did test it right now and it is working. Please merge both #410 and #411 if you don't see any further pending issues.
